sdks/python/apache_beam/runners/common.py

```diff
@@ -845,52 +880,241 @@ def _invoke_process_per_window(self,
-      current_watermark = (threadsafe_watermark_estimator.current_watermark())
-      estimator_state = (threadsafe_watermark_estimator.get_estimator_state())
-      split = threadsafe_restriction_tracker.try_split(fraction)
-      if split:
-        primary, residual = split
-        element = current_windowed_value.value
-        restriction_provider = self.signature.get_restriction_provider()
-        primary_size = restriction_provider.restriction_size(element, primary)
-        residual_size = restriction_provider.restriction_size(element, residual)
-        primary_value = ((element, (primary, None)), primary_size)
-        residual_value = ((element, (residual, estimator_state)), residual_size)
-        return (
-            SplitResultPrimary(
-                primary_value=current_windowed_value.with_value(primary_value)),
-            SplitResultResidual(
-                residual_value=current_windowed_value.with_value(
-                    residual_value),
-                current_watermark=current_watermark,
-                deferred_timestamp=None))
-    return None
+      result = PerWindowInvoker._try_split(
+          fraction,
+          self.current_window_index,
+          self.stop_window_index,
+          self.current_windowed_value,
+          self.restriction,
+          self.watermark_estimator_state,
+          self.signature.get_restriction_provider(),
+          self.threadsafe_restriction_tracker,
+          self.threadsafe_watermark_estimator)
+      if not result:
+        return None
+
+      residuals, primaries, self.stop_window_index = result
+      return (residuals, primaries)
+
+  @staticmethod
+  def _scale_progress(progress, window_index, stop_window_index):
+    # We scale progress based upon the amount of work we will do for one
+    # window and have it apply for all windows.
+    completed = window_index * progress.total_work + progress.completed_work
+    remaining = (
+        stop_window_index -
+        (window_index + 1)) * progress.total_work + progress.remaining_work
+    from apache_beam.io.iobase import RestrictionProgress
+    return RestrictionProgress(completed=completed, remaining=remaining)
 
   def current_element_progress(self):
     # type: () -> Optional[RestrictionProgress]
-    restriction_tracker = self.threadsafe_restriction_tracker
-    if restriction_tracker:
-      return restriction_tracker.current_progress()
-    else:
+    if not self.is_splittable:
       return None
 
+    with self.splitting_lock:
+      current_window_index = self.current_window_index
+      stop_window_index = self.stop_window_index
+      threadsafe_restriction_tracker = self.threadsafe_restriction_tracker
+
+    if not threadsafe_restriction_tracker:
+      return None
+
+    progress = threadsafe_restriction_tracker.current_progress()
+    if not current_window_index or not progress:
+      return progress
+
+    assert stop_window_index
```

Comment:
Can we add a comment here to explain why `stop_window` should not be `None`?

Reply:
Done